cpp/plugins/cucim.kit.cuslide2/src/cuslide/cuslide.cpp

```diff
@@ -0,0 +1,382 @@
+        }
+    }
+
+    //
+    // Metadata Setup
+    //
+
+    // Note: int-> uint16_t due to type differences between ImageMetadataDesc.ndim and DLTensor.ndim
+    const uint16_t ndim = 3;
+    auto& resource = out_metadata.get_resource();
+
+    std::string_view dims{ "YXC" };
+
+    const auto& level0_ifd = tif->level_ifd(0);
+    std::pmr::vector<int64_t> shape(
+        { level0_ifd->height(), level0_ifd->width(), level0_ifd->samples_per_pixel() }, &resource);
+
+    DLDataType dtype{ kDLUInt, 8, 1 };
+
+    // TODO: Fill correct values for cucim::io::format::ImageMetadataDesc
+    uint8_t n_ch = level0_ifd->samples_per_pixel();
+    if (n_ch != 3)
+    {
+        // Image loaded by a slow-path(libtiff) always will have 4 channel
+        // (by TIFFRGBAImageGet() method in libtiff)
+        n_ch = 4;
+        shape[2] = 4;
+    }
+    std::pmr::vector<std::string_view> channel_names(&resource);
+    channel_names.reserve(n_ch);
+    if (n_ch == 3)
+    {
+        channel_names.emplace_back(std::string_view{ "R" });
+        channel_names.emplace_back(std::string_view{ "G" });
+        channel_names.emplace_back(std::string_view{ "B" });
+    }
+    else
+    {
+        channel_names.emplace_back(std::string_view{ "R" });
+        channel_names.emplace_back(std::string_view{ "G" });
+        channel_names.emplace_back(std::string_view{ "B" });
+        channel_names.emplace_back(std::string_view{ "A" });
+    }
+
+    // Spacing units
+    std::pmr::vector<std::string_view> spacing_units(&resource);
+    spacing_units.reserve(ndim);
+
+    std::pmr::vector<float> spacing(&resource);
+    spacing.reserve(ndim);
+    const auto resolution_unit = level0_ifd->resolution_unit();
+    const auto x_resolution = level0_ifd->x_resolution();
+    const auto y_resolution = level0_ifd->y_resolution();
+
+    switch (resolution_unit)
+    {
+    case 1: // no absolute unit of measurement
+        spacing.emplace_back(y_resolution);
+        spacing.emplace_back(x_resolution);
+        spacing.emplace_back(1.0f);
+
+        spacing_units.emplace_back(std::string_view{ "" });
+        spacing_units.emplace_back(std::string_view{ "" });
+        break;
+    case 2: // inch
+        spacing.emplace_back(y_resolution != 0 ? 25400 / y_resolution : 1.0f);
+        spacing.emplace_back(x_resolution != 0 ? 25400 / x_resolution : 1.0f);
+        spacing.emplace_back(1.0f);
+
+        spacing_units.emplace_back(std::string_view{ "micrometer" });
+        spacing_units.emplace_back(std::string_view{ "micrometer" });
+        break;
+    case 3: // centimeter
+        spacing.emplace_back(y_resolution != 0 ? 10000 / y_resolution : 1.0f);
+        spacing.emplace_back(x_resolution != 0 ? 10000 / x_resolution : 1.0f);
+        spacing.emplace_back(1.0f);
+
+        spacing_units.emplace_back(std::string_view{ "micrometer" });
+        spacing_units.emplace_back(std::string_view{ "micrometer" });
```

Comment:
Same question about assigning global `const`s for `string_view`s and reusing them